### DISCUSSION TITLE
Task was destroyed but it is pending!

### DISCUSSION THREAD

**Discussion by euri10** [frankie567/httpx-ws#33] [Q&A]
is there a way to avoid those warnings on my tests ?

```
Task was destroyed but it is pending!
task: <Task pending name='Task-4' coro=<AsyncWebSocketSession._background_receive() done, defined at /home/lotso/.cache/pypoetry/virtualenvs/lemongpt-jx0mSZeq-py3.11/lib/python3.11/site-packages/httpx_ws/_api.py:915> wait_for=<Future finished result=None>>
Task was destroyed but it is pending!
task: <Task pending name='Task-5' coro=<AsyncWebSocketSession._background_keepalive_ping() done, defined at /home/lotso/.cache/pypoetry/virtualenvs/lemongpt-jx0mSZeq-py3.11/lib/python3.11/site-packages/httpx_ws/_api.py:952> wait_for=<Future pending cb=[Task.task_wakeup()]>>
Task was destroyed but it is pending!
task: <Task pending name='Task-9' coro=<sleep() done, defined at /home/lotso/.asdf/installs/python/3.11.2/lib/python3.11/asyncio/tasks.py:627> wait_for=<Future pending cb=[Task.task_wakeup()]>>
```

**Comment by frankie567** [comment]
Hmm 🤔 

In your test code, do you make sure the [`.close()`](https://frankie567.github.io/httpx-ws/reference/httpx_ws/#httpx_ws._api.AsyncWebSocketSession.close) method is called before ending the test session? Either implicitly by properly ending the context manager, or explicitly by calling `.close()`?

**Reply by euri10** [reply]
yup definitely:
```
@pytest.mark.asyncio
async def test_websocket(app: FastAPI) -> None:
    async with httpx.AsyncClient(
        transport=ASGIWebSocketTransport(app), base_url="http://testserver"
    ) as client:
        async with aconnect_ws(
            "/ws?temp=0.0&model=gpt-3.5-turbo&token=1024", client
        ) as ws:
            await ws.send_json(
                {"HEADERS": "htmx headers", "chat-message-name": "Hello"}
            )
            response = await ws.receive_text()
            logger.debug(f"response: {response}")
            assert (
                response
                == "<div id='notifications' hx-swap-oob='beforeend'><div>USER: Hello</div></div>"
            )
```

**Reply by euri10** [reply]
mmm I may have a hint, "issue" is solved if  I replace `@pytest.mark.asyncio` by

```@pytest.fixture
def anyio_backend():
    return 'asyncio'

@pytest.mark.anyio
```

**Reply by frankie567** [reply]
Huh, so it probably means bizarre things happen with the asyncio loop 😩

Out of curiosity, does it still happen if you add `await asyncio.sleep()` before the end of your test?

**Reply by euri10** [reply]
a `await asyncio.sleep(0.1)` (0.0 doesnt work, in fact any delay above 0.0 seems ok) works if placed either at same level of the first context manager or second one, it still warns me with the sleep inside at assert level inside the 2 contexts managers.

**Reply by euri10** [reply]
seems like https://github.com/pytest-dev/pytest-asyncio/pull/309 would solve it

**Reply by frankie567** [reply]
Oh, interesting! Thanks for the investigation @euri10 :) I'll watch this!

**Reply by euri10** [reply]
finally found the time and I can confirm the PR above fixes the warnings.
I also quickly watched over anyio marker implementation and didn't see an equivalent "cleaning" process though.

**Reply by frankie567** [reply]
Thanks 🙏
